processor/resourceprocessor/resource_processor.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/processor"
+)
+
+type resourceTraceProcessor struct {
+	resource     *resourcepb.Resource
+	capabilities processor.Capabilities
+	next         consumer.TraceConsumer
+}
+
+func newResourceTraceProcessor(next consumer.TraceConsumer, cfg *Config) *resourceTraceProcessor {
+	resource := createResource(cfg)
+	return &resourceTraceProcessor{
+		next:         next,
+		capabilities: processor.Capabilities{MutatesConsumedData: !isEmptyResource(resource)},
+		resource:     resource,
+	}
+}
+
+// ConsumeTraceData implements the TraceProcessor interface
+func (rtp *resourceTraceProcessor) ConsumeTraceData(ctx context.Context, td consumerdata.TraceData) error {
+	return rtp.next.ConsumeTraceData(ctx, consumerdata.TraceData{
+		Node:         td.Node,
+		Resource:     mergeResource(td.Resource, rtp.resource),
+		Spans:        td.Spans,
+		SourceFormat: td.SourceFormat,
+	})
+}
+
+// GetCapabilities returns the Capabilities assocciated with the resource processor.
+func (rtp *resourceTraceProcessor) GetCapabilities() processor.Capabilities {
+	return rtp.capabilities
+}
+
+// Shutdown is invoked during service shutdown.
+func (*resourceTraceProcessor) Shutdown() error {
+	return nil
+}
+
+type resourceMetricProcessor struct {
+	resource     *resourcepb.Resource
+	capabilities processor.Capabilities
+	next         consumer.MetricsConsumer
+}
+
+func newResourceMetricProcessor(next consumer.MetricsConsumer, cfg *Config) *resourceMetricProcessor {
+	resource := createResource(cfg)
+	return &resourceMetricProcessor{
+		resource:     resource,
+		capabilities: processor.Capabilities{MutatesConsumedData: !isEmptyResource(resource)},
+		next:         next,
+	}
+}
+
+// GetCapabilities returns the Capabilities assocciated with the resource processor.
+func (rmp *resourceMetricProcessor) GetCapabilities() processor.Capabilities {
+	return rmp.capabilities
+}
+
+// Shutdown is invoked during service shutdown.
+func (*resourceMetricProcessor) Shutdown() error {
+	return nil
+}
+
+// ConsumeMetricsData implements the MetricsProcessor interface
+func (rmp *resourceMetricProcessor) ConsumeMetricsData(ctx context.Context, md consumerdata.MetricsData) error {
+	mergeResource(md.Resource, rmp.resource)
+	return rmp.next.ConsumeMetricsData(ctx, consumerdata.MetricsData{
+		Node:     md.Node,
+		Resource: mergeResource(md.Resource, rmp.resource),
+		Metrics:  md.Metrics,
+	})
+}
+
+func createResource(cfg *Config) *resourcepb.Resource {
+	rpb := &resourcepb.Resource{
+		Type:   cfg.ResourceType,
+		Labels: map[string]string{},
+	}
+	for k, v := range cfg.Labels {
+		rpb.Labels[k] = v
+	}
+	return rpb
+}
+
+func mergeResource(to, from *resourcepb.Resource) *resourcepb.Resource {
+	if isEmptyResource(from) {
+		return to
+	}
+	if to == nil {
+		to = &resourcepb.Resource{Labels: map[string]string{}}
```

Comment:
just being paranoid here: it could be that `to` is not nil but `Labels` is ... in principle I don't see a case for that but seems worth to prevent a panic since we will be writing to the map.

Reply:
Done